### DISCUSSION TITLE
What happened to Glint? (lsp)

### DISCUSSION THREAD

**Discussion by NullVoxPopuli** [mason-org/mason.nvim#583] [Q&A]
This seems to have been missed in migration or something? or maybe I'm behind -- feels like a recently occurring issue, but I'm getting this error:
```
[mason-lspconfig.nvim] Server "glint" is not a valid entry in ensure_installed. Make sure to only provide lspconfig server names.
```        

When I change `glint` to `glint-language-server`, I get:
```
[mason-lspconfig.nvim] Server "glint-language-server" is not a valid entry in ensure_installed. Make sure to only provide lspconfig server names.
[lspconfig] Cannot access configuration for glint-language-server. Ensure this server is listed in `server_configurations.md` or added as a custom server.
```

My LSP Packer setup: https://github.com/NullVoxPopuli/dotfiles/blob/main/home/.config/nvim/lua/plugins/lsp.lua

<details><summary>at the time of posting</summary>

```lua
return function(use)
  use 'neovim/nvim-lspconfig'
  use { 'williamboman/mason.nvim' }
  use { 'williamboman/mason-lspconfig.nvim' }

  -- Development
  -- use '~/Development/NullVoxPopuli/nvim-lspconfig'
  -- use { 'muziejus/nvim-lspconfig', branch = "glint" }

  -- Completion
  use {
    'hrsh7th/nvim-cmp',
     requires = {
       -- Completion Sources
       'hrsh7th/cmp-nvim-lsp',
       -- Considered hacky (by the author)
       -- https://github.com/hrsh7th/cmp-nvim-lsp-signature-help/issues/12#issuecomment-1128820104
       -- 'hrsh7th/cmp-nvim-lsp-signature-help',
       'hrsh7th/cmp-buffer',
       'hrsh7th/cmp-path',
       'hrsh7th/cmp-nvim-lua',
       -- 'f3fora/cmp-spell',

       -- It bothers me that this person's name is all caps
       'L3MON4D3/LuaSnip'
      }
    }

  use 'ray-x/lsp_signature.nvim'


  -- LSP status on the bottom right as hint text.
  -- subtle.
  --  Better than:
  --    use 'arkav/lualine-lsp-progress'
  --    use 'nvim-lua/lsp-status.nvim'
  --    use 'doums/lsp_spinner.nvim'
  --
  --  In that, this works (and is unobtrusive visually, and to the config)
  use 'j-hui/fidget.nvim'

  use {
    "smjonas/inc-rename.nvim",
    config = function()
      -- Default: <leader>r
      require("inc_rename").setup()
    end,
  }
end

```

</details>

My config: https://github.com/NullVoxPopuli/dotfiles/blob/main/home/.config/nvim/lua/plugin-config/lsp.lua#L25

<details><summary>at the time of posting</summary>

```lua
local cmp = require'cmp'
local lsp = require('lspconfig')
local notify = require('notify')
-- local lsp_status = require('lsp-status')
-- local lsp_spinner = require('lsp_spinner')

local servers = {
  --------------
  -- Languages
  "html",
  "jsonls",
  "yamlls",
  "cssls",
  "sumneko_lua",
  "tsserver",
  "bashls",
  -- Not actively using these atm
  -- "elixirls",
  -- "rust_analyzer",
  -- "fsautocomplete",

  --------------
  -- Frameworks
  "ember",
  "glint",

  --------------
  -- Tools
  "graphql",
  "tailwindcss",
  "graphql",
  "dockerls",

  --------------
  -- Linting
  "eslint",
}

vim.keymap.set("n", "<A-r>", ":IncRename ")

local mySettings = {
  eslint = {
    format = true,
  },
  -- some config ommitted
  tsserver = {
    maxTsServerMemory = 8000,
    implicitProjectConfig = {
      experimentalDecorators = true
    },
    -- importModuleSpecifier = "shortest"
  }
}

-- https://github.com/j-hui/fidget.nvim
require"fidget".setup {}

require("mason").setup {
  ui = {
    icons = {
      server_installed = "✓",
      server_pending = "➜",
      server_uninstalled = "✗"
    }
  }
}
require("mason-lspconfig").setup {
  ensure_installed = servers,
  automatic_installation = true
}

-- etc -- the rest is omitted --
```

</details>

in `lsp-config`, the Glint server configuration is here: https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/server_configurations/glint.lua

**Comment by williamboman** [comment]
Hey! Hm to my knowledge (and according to git history) `glint` has never been a package provided by Mason. That warning message was added recently to notify users if they're providing servers that don't exist to `ensure_installed` (which basically is a noop).

**Reply by NullVoxPopuli** [reply]
since `glint` exists in lsp-config, what's the path forward? I'm not totally familiar with these two plugins interact _exactly_, but I am willing to put up a PR, if I know what needs to be done.

Does`glint` or `glint-language-server` need to be added to this list of things?: https://github.com/williamboman/mason-lspconfig.nvim/tree/main/lua/mason-lspconfig/server_configurations -- but why wouldn't `tsserver` have similar warnings :thinking:

**Reply by williamboman** [reply]
The directory you linked to are only for servers that require additional configuration to be usable. Generally this is not needed which is why that list is short.

First a Mason package would need to [be added to Mason itself](https://github.com/williamboman/mason.nvim/blob/main/CONTRIBUTING.md). Next, a mapping entry [needs to be added in `mason-lspconfig`](https://github.com/williamboman/mason-lspconfig.nvim/blob/main/lua/mason-lspconfig/mappings/server.lua) + an entry in the [README](https://github.com/williamboman/mason-lspconfig.nvim#available-lsp-servers).

**Reply by NullVoxPopuli** [reply]
Thanks! I think we're good to go:
- https://github.com/williamboman/mason.nvim/pull/593
- https://github.com/williamboman/mason-lspconfig.nvim/pull/89

Local testing in my glint-using monorepeo seemed good.
And it seems to be a better experience than before -- which was that glint _and_ the tsserver started at the same time (which isn't what we want since glint wraps tsserver, and provides glimmer-specific augmentations) -- and now only Glint starts for my glint project. and TSserver still works on my ts-only files. legit.

One thing I did notice that I'm not sure what to do about is that my LSP-status plugin (for showing when an LSP server is booting) doesn't work with Glint -- glint just silently starts after < couple seconds. 
I'm using https://github.com/j-hui/fidget.nvim/ -- I need to do some digging in there to see how that works

**Reply by NullVoxPopuli** [reply]
Well, maybe I should wait for the author to respond to this first -- https://github.com/j-hui/fidget.nvim/issues/107

Maybe a bunch of my LSPs accidentally work somehow

**Reply by williamboman** [reply]
There's nothing Mason does that could interfere the [`$/progress`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#progress) events. Are you certain the glint language server emits these events?

**Reply by NullVoxPopuli** [reply]
I am not -- I haven't actually worked on the language server itself -- I will investigate
